### ISSUE TITLE
[Tune] Making my stopper more efficient

### ISSUE THREAD

**Issue by flokde** [ray-project/ray#10101]
[tune]
I have an objective function (Trainable) to optimise through hyperopt, that does not profit from additional iterations. The optimization is time constrained. So I want to draw samples from the search space and evaluate them until this time is used up. To implement this, I create Trainables and only run them for one iteration (this was the best solution I could come up with). I defined a custom stopper for this purpose:

```
class SingleTrialTimeStopper(tune.Stopper):
    def __init__(self, time_limit):
        self._start = time.time()
        self._deadline = time_limit

    def __call__(self, trial_id, result):
        if result['training_iteration'] >= 1:
            should_stop = True
        else:
            should_stop = False
        return should_stop

    def stop_all(self):
        return time.time() - self._start > self._deadline
```

I assume this is a fairly standard case and have the suspicion my stopper is unnecessarily complicated. When running the optimization, once the time limit is reached, I repeatedly get this warning:

`WARNING ray_trial_executor.py:84 -- Skipping cleanup - trainable.stop did not return in time. Consider making stop a faster operation.`

This happens every minute (sometimes for over 30mins in total). Is there a better way to define a stopper for my use case in order to avoid this? Help would be much appreciated!

*Ray 0.8.6, Python 3.6.8, CentOS Version 7 (Core)*

**Comment by rkooo567** [comment]
cc @richardliaw

**Comment by richardliaw** [comment]
Hmm, in your trainable, you can do:

```
result_dict = ...
result_dict["done"] = True
return result_dict (or if using the function API, tune.report(**result_dict)
```

and then you wouldn't need to implement `__call__` in your stopper. 
Also, try implementing `Trainable.reset_config`, assuming you're using the class API?

**Comment by flokde** [comment]
Hi Richard,

thank you so much for your answer! So, solution 1 pretty much does what I am doing already, but more efficient if I understand correctly?

I am really interested In implementing your second solution - I am indeed using class API. This would be super elegant, as I only need to create one Trainable and then use iterations on it. My one concern (why I also decided against it before) is the use of an optimiziation algo. 

I am using hyperopt and as per this issue [8674](https://github.com/ray-project/ray/issues/8674) my understanding was I have to generate new Trainables to actually gain benefit from the optimisation through hyperopt. But I have to admit I couldn't figure out from the documentation/code if this would maybe also work when using `Trainable.reset_config`. I also commented on that issue. Basically I would really like to use `Trainable.reset_config` as it is much cleaner, but am unsure if it would work with hyperopt. If you could help me understand this, I would be grateful.

Thank you very much for your help!

**Comment by richardliaw** [comment]
Hey @flokde, no problem! 

> thank you so much for your answer! So, solution 1 pretty much does what I am doing already, but more efficient if I understand correctly?

yep! 

> I am really interested In implementing your second solution - I am indeed using class API. This would be super elegant, as I only need to create one Trainable and then use iterations on it. My one concern (why I also decided against it before) is the use of an optimiziation algo.
> 
> I am using hyperopt and as per this issue [8674](https://github.com/ray-project/ray/issues/8674) my understanding was I have to generate new Trainables to actually gain benefit from the optimisation through hyperopt. But I have to admit I couldn't figure out from the documentation/code if this would maybe also work when using `Trainable.reset_config`. I also commented on that issue. Basically I would really like to use `Trainable.reset_config` as it is much cleaner, but am unsure if it would work with hyperopt. If you could help me understand this, I would be grateful.

Ah I see. It should be fine. I think all you need to understand is the general lifecycle of a Trainable:

```
new_trainable = Trainable.setup()
result = new_trainable.step()
assert "done" in result # this is your code

new_trainable.reset_config(new_config)
result = new_trainable.step()
assert "done" in result # this is your code

...
# at end of experiment
new_trainable.stop()
```
Basically, you will want to implement `reset_config` such that it creates a new instance of the function you're optimizing. This can be tricky if you're using something very custom with Tensorflow + CUDA drivers, as it is hard to reset the training environment. But I think for your case, it sounds like it should work out of the box!

**Comment by flokde** [comment]
Your last answer really helped me understand ray/tune a little bit better. Currently I am not getting there on the implementation of `reset_config(config)`, so I am abusing this issue a little bit - I assume you will stop answering my question if they get too dumb :).

I went for the most basic `reset_config` method possible:
```
def reset_config(self, new_config):

    self.config = utils.adapt_parameter_config(new_config,
                                               type(self).optimizer,
                                               type(self).min_count,
                                               type(self).min_maj_ratio)
    return True
```

When I go into debugging and call the method (with a dummy dict as input) this seems to work. However, the method is never called when I use `tune.run` -  I checked this with inserting a trace and print statements. My assumption was that the lifecycle of a Trainable is managed through `tune.run`. E.g. the `_setup` (now `setup` in 0.8.7) method works nicely without me having to call it somewhere. `tune.run` takes care of pulling a parameter configuration from the hyperopt search space and passes it to `_setup(config)`, so I thought `reset_config(config)` would do the same. Do I have to manually call `reset_config` somewhere?

My `tune.run` looks like this:
```
hyperopt_algo = HyperOptSearch(self.search_space,
                               metric="score",
                               mode=self.mode,
                               n_initial_points=16)

hyperopt_algo = ConcurrencyLimiter(hyperopt_algo, max_concurrent=4)
  
self.tune_analysis = tune.run(
               Trainable,
               config=self.search_space,
               search_alg=hyperopt_algo,
               resources_per_trial={"cpu": self.number_of_cores},
               stop=SimpleTimeStopper(time_limit=self.time_limit_seconds),
               num_samples=self.num_samples,
               local_dir=self.out_dir,
               raise_on_failed_trial=False,
               reuse_actors=True,
               ray_auto_init=True)
```

I went through old issues and eventually the code, but it is too advanced for me at points, so I didn't really find the point, where `reset_config` is called. Again, thank you very much your help.

All the best!

**Comment by richardliaw** [comment]
Hmm, actually it looks like for your case reuse_actors wouldn't work. Sorry about that! How long does each function evaluation take ('Trainable.step') take?

If it is very fast, consider running `ray.init(local_mode=True)` before `tune.run`. This will force everything to be a single-process implementation.

Alternatively, if it is slow, you might want to consider taking advantage of parallelism. `resources_per_trial={"cpu": self.number_of_cores},` means that each single evaluation will reserve `self.number_of_cores` cores. If I understand your code correctly, where self.number_of_cores == total cores, then it seems like you won't be able to take advantage of the parallelism that Ray provides.

**Comment by flokde** [comment]
Ah that is too bad. I had some problems with memory and I assumed it was because I saved the data in my Trainable subclass. So reusing actors wouldn't only have been elegant and more efficient with the stoppers, but might have also helped with the memory problems. The step/train time depends a lot on what kind of data I input - but for the most important case it is generally between 15sec to 3min depending on the hyperparameters (heavy downsampling can be involved).

Thanks for your other feedback. I am calling` ray.init `with local mode before `tune.run` anyways (usually have ray_auto_init=False, just thought it might be confusing and as I am writing this I realise I should have just left the line out), as I try to restrict the memory through this - with medium success. And my code is not super clean, but essentially parallelisation is what I am doing with self.number_of_cores. I am assigning 2 cores per Trial to utilise parallelisation. But you couldn't possibly know this from my non-clean code snippet and missing the context :).

Thanks for your help, I learned a lot and will definitely come back to ray in the future!

**Comment by richardliaw** [comment]
Got it! BTW @flokde what type of data are you saving and how is it affecting your memory usage?

Maybe there's a way to resolve that issue!

**Comment by flokde** [comment]
Thanks so much, I would be excited to learn more about the framework and understand it better, so I'll try to outline my problem here and if you have the time and get around to having a look and giving me feedback, that would be great! 

I am working on finding optimal sklearn pipelines with a couple of additional algorithms and preprocessing operations added to the search space. I want it to be like a minimal framework, so the user can put in their data of choice. My go to use case is a large tabular dataset (between 10k and 10mio datapoint - though the bigger datasets get downsampled, so realistically up to 300k).

As the objective function depends on the data and some characteristics of the dataset, I can't just define a subclass `PipelineTrainable` of `Trainable`, because it differs from dataset to dataset. I ended up solving that by introducing yet another subclass. 

I basically define a subclass `PipelineTrainable`, that inherits from `Trainable` that is the outline of my objective function (but still needs the data-specific characteristics to work) - so far so good/normal. Now for the "weirder" part:

In my mini-framework I have a fit method, that runs `tune.run` to find the optimal pipeline. Before calling` tune.run` I create a subclass `CustomPipelineTrainable` that inherits from `PipelineTrainable` that gets passed the dataset and a few characteristics as class attributes. My fear is that somehow with each Trainable that gets started (which can be hundreds or thousands, depending on the size of the dataset) somehow a copy of the dataset is stored somewhere. When I run my fit method the memory just keeps going up until it crashes. I tried combatting this the following ways:

1. I know ray does garbage collect, but I tried calling gc.collect() manually in my Trainable.

2. I used `ray.init` before `tune.run` and disabled `ray_auto_init`:
`ray.init(memory=10*1024*1024*1024, object_store_memory=2*1024*1024*1024, redis_max_memory=8*1024*1024*1024, driver_object_store_memory=5*1024*1024*1024, lru_evict=False)`
This is for a 32GB machine, that is solely for running my framework. I also tried `lru_evict=True`.

3. I overwrote the deep copy method of my sub-subclass and added a print statement to see if the deep copy method would be called (again, because I thought that with each Trainable a copy is made and thus the dataset is copied once again). Maybe now you understand, why I was so excited about reusing actors :).

None of this really helped and I didn't know where else to go from there. I ended up just putting a pin in it and upped my memory for now.

I hope this wasn't too prolonged/confusing. I would really like to understand where I went wrong or what I am missing. I also have the suspicion that the sub-subclass is not the most elegant solution.

Again, thanks for you help and patience!

**Comment by richardliaw** [comment]
Ok no this is really helpful. Have you taken a look at tune-sklearn? https://github.com/ray-project/tune-sklearn

It provides a scikit-learn interface and handles large datasets properly underneath the hood. Would it work for your particular use case? (You’ll have to install tune-sklearn from source if you try it out).

**Comment by richardliaw** [comment]
The main thing we do in tune-sklearn to address large datasets is that we leverage the Ray object store:

x_id = ray.put(dataset)

and on the trainable

dataset = ray.get(x_id)

This allows us to leverage shared memory, only saving one copy of the data across all trainables throughout the lifetime of the experiment.
